### PR TITLE
[azure-service-bus] Add possibility to receive messages from Azur…

### PR DESCRIPTION
**Pull request by vkepin**
…e Service Bus

### CODE REVIEW

File: vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/model/ServiceBusClientEntity.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package org.vividus.azure.servicebus.model;
+
+public class ServiceBusClientEntity
```
Comment (vkepin) [inline vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/model/ServiceBusClientEntity.java:3]: remove 'builder' style

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 2]
```diff
@@ -13,25 +13,122 @@ include::partial$azure-authentication.adoc[]
...
+
+* `$serviceBusKey` - The key associated with the Azure Service Bus client, will be used to find required properties values.
+
+.Start consuming messages from the topic configured by Service Bus key `myProjectTopic`
+[source,gherkin]
+----
+When I start consuming messages from `myProjectTopic` service bus
+----
+
+=== Stop consume messages
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:75]:
```suggestion
=== Stop consuming messages
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 3]
```diff
@@ -13,25 +13,122 @@ include::partial$azure-authentication.adoc[]
...
+----
+
+=== Save messages to the variable
+
+Drains/Peeks the consumed messages to the specified variable. If the consumer is not stopped, the new events might arrive after the draining.
+If the consumer is stopped, all the events received from the consumer start or after the last draining operation are stored to the variable.
+
+[source,gherkin]
+----
+When I $queueOperation consumed `$serviceBusKey` service bus messages to $scopes variable `$variableName`
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:121]: I don't think it's for queue only

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 4]
```diff
@@ -13,25 +13,122 @@ include::partial$azure-authentication.adoc[]
...
+* `$queueOperation` - `DRAIN` - saves the messages consumed since the last drain or from the consumption start and moves the consumer cursor to the position after the last consumed message;
+`PEEK` - saves the messages consumed since the last drain or from the consumption start and doesn't change the consumer cursor position.
+* `$serviceBusKey` - The key associated with the Azure Service Bus client, will be used to find required properties values.
+* `$scopes` - xref:commons:variables.adoc#_scopes[The comma-separated set of the variables scopes].
+* `$variableName` - The variable name to store the messages. The messages are accessible via zero-based index, e.g. `${my-var[0]}` will return the first received message.
+
+.Peek messages from the consumer and save to variable
+[source,gherkin]
+----
+When I PEEK consumed `myProjectQueue` service bus messages to STORY variable `queueMessages`
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:133]: please add a set of complete examples to get understanding how to use these steps (in the same way we have for Kafka)

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 5]
```diff
@@ -13,25 +13,122 @@ include::partial$azure-authentication.adoc[]
...
+Drains/Peeks the consumed messages to the specified variable. If the consumer is not stopped, the new events might arrive after the draining.
+If the consumer is stopped, all the events received from the consumer start or after the last draining operation are stored to the variable.
+
+[source,gherkin]
+----
+When I $queueOperation consumed `$serviceBusKey` service bus messages to $scopes variable `$variableName`
+----
+
+* `$queueOperation` - `DRAIN` - saves the messages consumed since the last drain or from the consumption start and moves the consumer cursor to the position after the last consumed message;
+`PEEK` - saves the messages consumed since the last drain or from the consumption start and doesn't change the consumer cursor position.
```
Comment (uarlouski) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:125]: does it makes sense to represent operations as sublist of queueOperation?

File: vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java [hunk 6]
```diff
@@ -17,57 +17,153 @@
...
-            LOGGER.info("The message was successfully sent to {} in {} namespace", name, namespaceName);
+            LOGGER.info("The message was successfully sent to {} in {} namespace", name, namespace);
         }
         catch (ServiceBusException e)
         {
             throw new ServiceBusSendMessageException("Unable to send message to Azure Service Bus", e);
         }
     }
+
+    public void startConsuming(String clientKey)
```
Comment (uarlouski) [inline vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java:99]: do we need to handle cases when end user invokes startConsuming several times with the same key?

File: vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java [hunk 7]
```diff
@@ -17,57 +17,153 @@
...
+        else
+        {
+            throw new IllegalArgumentException(
+                    "There are no running Azure Service Bus consumers associated with the key:" + clientKey);
+        }
+    }
+
+    public void stopAll()
+    {
+        clients.keySet().stream().toList().forEach(this::stopConsuming);
```
Comment (uarlouski) [inline vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java:161]: forEach can be invoked on stream

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 8]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+
+[source,properties]
+----
+azure.service-bus.{service-bus-key}.{property-name}=property value
+----
+
+where:
+
+. `service-bus-key` - The key associated with the Azure Service Bus configuration, which will be used as a step parameter.
+. `property-name` - The name of the topic property. One of:
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:28]:
```suggestion
. `property-name` - The name of Service Bus property. One of:
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 9]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+.. `channel-type` - The type of service bus messaging components: either
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:29]:
```suggestion
.. `channel-type` - The type of Service Bus messaging channel: either
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 10]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-messaging-overview#concepts[QUEUE or TOPIC].
+.. `namespace` - The name of the namespace the service bus belongs to.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:31]:
```suggestion
.. `namespace` - The name of the namespace Service Bus belongs to.
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 11]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+.. `name` - The queue or topic name.
+.. `subscription-name` - The name of the topic subscription. Only for TOPIC channel-type.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:33]:
```suggestion
.. `subscription-name` - The name of the topic subscription. Only for `TOPIC` channel type.
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 12]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
 .Send simple message to the Service Bus Queue
 [source,gherkin]
 ----
-When I send message to service bus QUEUE with name `testQueue` in namespace `myProjectNamespace` and payload:`Hello World!`
+When I send message to `myProjectQueue` service bus with payload:`Hello World!`
+----
+
+=== Start consuming messages
+
+Starts the Azure Service Bus consumer with the provided configuration to listen the specified
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:57]:
```suggestion
Starts Azure Service Bus consumer with the provided configuration to listen the specified
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 13]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-messaging-overview#concepts[Queue or Topic].
+
+IMPORTANT: The consumer must be stopped when it's not needed.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:60]:
```suggestion
IMPORTANT: The consumer must be stopped when it's not used.
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 14]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+
+.Start consuming messages from the topic configured by Service Bus key `myProjectTopic`
+[source,gherkin]
+----
+When I start consuming messages from `myProjectTopic` service bus
+----
+
+=== Stop consuming messages
+
+Stops the Azure Service Bus consumer started by the <<_start_consuming_messages>> step before.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:77]:
```suggestion
Stops Azure Service Bus consumer started by the <<_start_consuming_messages>> step before.
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 15]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+
+Waits until the count of the consumed messages (from the consumer start or after the last draining operation) matches to the rule or until the timeout is exceeded.
+
+[source,gherkin]
+----
+When I wait with `$timeout` timeout until count of consumed by `$serviceBusKey` service bus messages is $comparisonRule `$expectedCount`
+----
+
+* `$timeout` - The maximum time to wait for the events in {durations-format-link} format.
+* `$serviceBusKey` - The key associated with the Azure Service Bus configuration, will be used to find required properties values.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:104]:
```suggestion
* `$serviceBusKey` - The key associated with Azure Service Bus configuration, will be used to find required properties values.
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 16]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+* `$comparisonRule` - xref:parameters:comparison-rule.adoc[The comparison rule].
+* `$expectedCount` - The expected count of the events to be matched by the rule.
+
+.Wait for at least 5 messages from the queue for 30 seconds
+[source,gherkin]
+----
+When I wait with `PT30S` timeout until count of consumed by `myProjectQueue` service bus messages is >= `5`
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:111]:
```suggestion
When I wait with `PT30S` timeout until count of consumed `myProjectQueue` service bus messages is >= `5`
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 17]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+If the consumer is stopped, all the events received from the consumer start or after the last draining operation are stored to the variable.
+
+[source,gherkin]
+----
+When I $operation consumed `$serviceBusKey` service bus messages to $scopes variable `$variableName`
+----
+
+* `$operation` - `DRAIN` - saves the messages consumed since the last drain or from the consumption start and moves the consumer cursor to the position after the last consumed message;
+`PEEK` - saves the messages consumed since the last drain or from the consumption start and doesn't change the consumer cursor position.
+* `$serviceBusKey` - The key associated with the Azure Service Bus configuration, will be used to find required properties values.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:126]:
```suggestion
* `$serviceBusKey` - The key associated with Azure Service Bus configuration, will be used to find required properties values.
```

File: docs/modules/plugins/pages/plugin-azure-service-bus.adoc [hunk 18]
```diff
@@ -13,25 +13,127 @@ include::partial$azure-authentication.adoc[]
...
+* `$scopes` - xref:commons:variables.adoc#_scopes[The comma-separated set of the variables scopes].
+* `$variableName` - The variable name to store the messages. The messages are accessible via zero-based index, e.g. `${my-var[0]}` will return the first received message.
+
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-azure-service-bus.adoc:129]: == Examples

File: docs/modules/plugins/pages/plugin-kafka.adoc [hunk 19]
```diff
@@ -23,7 +23,7 @@ NOTE: The properties marked with *bold* are mandatory.
 |A comma-separated list of host and port pairs that are the addresses of the Kafka brokers in a "bootstrap" Kafka cluster that a Kafka client connects to initially to bootstrap itself.
 
 |===
-
+0
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-kafka.adoc:26]: 🗑️

File: vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/ServiceBusSteps.java [hunk 20]
```diff
@@ -16,30 +16,234 @@
...
+     * </code>
+     * <br>
+     * <br> Where:
+     * <ul>
+     *     <li><b>channel-type</b> - the type of service bus messaging components: either `QUEUE` or `TOPIC`</li>
+     *     <li><b>namespace</b> - the name of the namespace the service bus belongs to</li>
+     *     <li><b>name</b> - the queue or topic name</li>
+     *     <li><b>subscription-name</b> - the name of the topic subscription. Only for TOPIC channel-type</li>
+     * </ul>
+     * @param serviceBusKey Identifying the Service Bus connection details.
```
Comment (valfirst) [inline vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/ServiceBusSteps.java:70]: make sure to keep docs and javadocs aligned

File: vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/ServiceBusSteps.java [hunk 21]
```diff
@@ -16,30 +16,234 @@
...
+     */
+    @When("I wait with `$timeout` timeout until count of consumed by `$serviceBusKey` service bus messages is"
+            + " $comparisonRule `$expectedCount`")
+    public void waitForServiceBusMessages(Duration timeout, String serviceBusKey, ComparisonRule comparisonRule,
+                                          int expectedCount)
+    {
+        Matcher<Integer> countMatcher = comparisonRule.getComparisonRule(expectedCount);
+        Integer result = new DurationBasedWaiter(timeout, Duration.ofSeconds(1)).wait(
+                () -> serviceBusService.getMessagesForClient(serviceBusKey).size(), countMatcher::matches);
+        softAssert.assertThat("Total count of messages for Service Bus client with key: " + serviceBusKey,
```
Comment (valfirst) [inline vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/ServiceBusSteps.java:182]: please avoid using `client` naming

File: vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/model/ServiceBusClientEntity.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.azure.servicebus.model;
+
+public class ServiceBusClientEntity
```
Comment (valfirst) [inline vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/model/ServiceBusClientEntity.java:19]:
```suggestion
public class ServiceBusConnectionParameters
```

File: vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java [hunk 23]
```diff
@@ -17,57 +17,153 @@
...
+        ChannelType channelType = serviceBusClientEntity.getChannelType();
+        String name = serviceBusClientEntity.getName();
+        if (channelType == ChannelType.QUEUE)
+        {
+            processorBuilder.queueName(name);
+        }
+        else
+        {
+            processorBuilder.topicName(name);
+            processorBuilder.subscriptionName(serviceBusClientEntity.getSubscriptionName());
```
Comment (valfirst) [inline vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java:129]: we need to throw an error if subscription name is configured for queue

File: vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java [hunk 24]
```diff
@@ -17,57 +17,153 @@
...
+
+        clients.put(clientKey, client);
+        LOGGER.info("'{}' Azure Service Bus Consumer of messages from the {} '{}' in namespace {} is started",
+                clientKey, channelType, name, namespace);
+        client.start();
+    }
+
+    private ServiceBusClientEntity getServiceBusClientConfig(String clientKey)
+    {
+        return clientConfigs.get(clientKey, "No Service Bus client with key `%s` is configured", clientKey);
```
Comment (valfirst) [inline vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java:141]:
```suggestion
        return clientConfigs.get(clientKey, "No Service Bus connection with key `%s` is configured", clientKey);
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java (hunk 7 above)
Comment (vkepin) [reply vividus-plugin-azure-service-bus/src/main/java/org/vividus/azure/servicebus/service/ServiceBusService.java:166]: I use remove method over the clients map. Usage of forEach will lead to ConcurrentModificationException

**Review by uarlouski — APPROVED** [review]

**Pull request merged**